### PR TITLE
[proxy] Pre-flight requests to server /api/* endpoints

### PR DESCRIPTION
**Pull request by andrew-farries**
## Description

Before handling `server` `/api` requests, make a 'pre-flight' request to `/api/feature-flags/slow-database` (on `server`) and use the `X-Gitpod-Slow-Database` header in the response to decide where to send the actual request; either to the regular `server` service or the new `slow-server` service.

This allows us to use a feature flag to simulate the performance of these endpoints when they use a higher latency database connection.

## Related Issue(s)
Part of #9198 

Closes https://github.com/gitpod-io/gitpod/issues/14960.

## How to test

0. Log in to the preview environment.
1. Hit https://af-proxy-a5b2f4cfb56.preview.gitpod-dev.com/api/version and inspect the response timings. 
2. Change the targeting rule for the `slow_database` feature flag (non-production environment) to this (replacing the user id with your user id in the preview):
image

3. Wait up to 3 minutes for the configcat client to refresh the feature flag values.
4. Hit https://af-proxy-a5b2f4cfb56.preview.gitpod-dev.com/api/version and inspect the response timings.

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-slow-database
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/proxy/conf/Caddyfile
Comment (andrew-farries) [inline components/proxy/conf/Caddyfile:263]: I'd like to be able to replace the body of these `reverse_proxy` directives with a snippet to reduce duplication but ran into this Caddy bug with nested imports: https://github.com/caddyserver/caddy/issues/4914

Comment (andrew-farries) [reply components/proxy/conf/Caddyfile:263]: Will take a look at `forward_auth`. When I looked before it didn't seem to match our use case, but will look again 👀

Comment (andrew-farries) [reply components/proxy/conf/Caddyfile:263]: > Would it work do do something like:
> ...

This looks like it should work according the Caddy docs on [intercepting responses](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#intercepting-responses):

I tried a simplified version in this Gist:

https://gist.github.com/andrew-farries/a2ac60d51e6c812ded573720c1ffabf6

but the fallthrough case doesn't work. Caddy docs say:

> When a response handler is invoked, the response from the backend is not written to the client, and the configured handle_response route will be executed instead, and it is up to that route to write a response. If the route does not write a response, then request handling will continue with any handlers that are ordered after this reverse_proxy.

but despite the `@fast` and `@none` matchers not writing a response, `respond` directive after the `reverse_proxy` directive never gets run.

(You can try it for yourself with `caddy run --watch` and then `http :2015/api/foo`)

Comment (andrew-farries) [reply components/proxy/conf/Caddyfile:263]: Another alternative to reduce duplication here would be to have the `feature-flags/slow-database` endpoint always set the `X-Gitpod-Slow-Database` response header. It currently only does so on `200 OK` responses:

https://github.com/gitpod-io/gitpod/blob/c154c3262a03a6b78ec40f8d07ef4e862cc9a52c/components/server/src/feature-flag/featureflag-controller.ts#L24-L44

That way we could remove the `@headerNotPresent` response matcher.

Comment (andrew-farries) [reply components/proxy/conf/Caddyfile]: Changed to use `forward_auth`, `copy_headers` and request matchers as we discussed in sync today. 

This needs https://github.com/gitpod-io/gitpod/pull/14930 to land first to ensure the `api/feature-flags/slow-database` endpoint returns `200 OK` consistently.

File: components/proxy/conf/Caddyfile
Comment (geropl) [inline components/proxy/conf/Caddyfile:247]:
```suggestion
		@fast not @slow
```
Is a tad shorter.

Also, we could move those matchers one level up, so we can re-use them in other handlers (but I think that is a follow-up). :100:

**Comment by geropl** [comment]
Currently I get an "Unauthorized", and figured that this needs to be rebased on https://github.com/gitpod-io/gitpod/pull/14930 (as you mentioned above :laughing: )

File: components/proxy/conf/Caddyfile
Comment (andrew-farries) [reply components/proxy/conf/Caddyfile:247]: Unfortunately, negating a named matcher with `not` isn't possible: https://github.com/caddyserver/caddy/issues/4112. So I think we are stuck with the more verbose version.

I'll move the matchers in a followup PR when we re-use them in other `handle` blocks.

**Comment by andrew-farries** [comment]
This has been rebased now that #14930 has been merged. Do you want to try again @geropl?

It is working as expected for me.

**Comment by geropl** [comment]
`{"@type":"type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent","serviceContext":{"service":"server","version":"af-proxy-api-to-slow-server.29"},"component":"server","severity":"WARNING","time":"2022-11-25T12:43:39.445Z","message":"websocket path not matching","payload":{}}`

I see a lot of these lines being logged, bc we forward websocket Upgrade-requests to the server, but on a rewritten URi.
I think a simple `header_up -Upgrade` inside of `forward_auth` would fix this. :+1:

**Review by geropl — APPROVED** [review]
And it works - sloooooowly :laughing: 

Nice work, @andrew-farries ! :rocket: 

/hold bc I think we should adress https://github.com/gitpod-io/gitpod/pull/14897#issuecomment-1327435344 before we merge

**Comment by andrew-farries** [comment]
> see a lot of these lines being logged, bc we forward websocket Upgrade-requests to the server, but on a rewritten URi.
I think a simple header_up -Upgrade inside of forward_auth would fix this. 👍

I'm not sure why we see these errors at all. In the `Caddyfile` we define two `handle` blocks, one with a path matcher for `api/gitpod`:

https://github.com/gitpod-io/gitpod/blob/54d7c00c88309d68f4e0566ba9639f8a25f37af4/components/proxy/conf/Caddyfile#L207-L230

and the other with a path matcher for `api/* headless-logs/*`:

https://github.com/gitpod-io/gitpod/blob/54d7c00c88309d68f4e0566ba9639f8a25f37af4/components/proxy/conf/Caddyfile#L232-L270

AIUI The `api/gitpod` path is the destination for all websocket requests, and as it is a more specific path than `api/*`, the `/api/gitpod` `handle` block will be the one used for incoming `Upgrade` requests [(see Caddy docs for 'handle')](https://caddyserver.com/docs/caddyfile/directives/handle). So why should anything we add to the `handle` block for `api/*` have an effect on websocket connections?

**Comment by andrew-farries** [comment]
After a sync today we determined that the cause of the websocket errors in the server log was websocket requests arriving at `api/v1` not being handled by the `handle @backend_wss` block but falling through to the modified `/api/*` handler block.

https://github.com/gitpod-io/gitpod/pull/14974 should resolve this.

**Comment by andrew-farries** [comment]
Using the version of `proxy` built by https://github.com/gitpod-io/gitpod/pull/14974 shows that the websocket errors are no longer present in the `server` logs 👍 .

**Pull request merged**
